scripts/scraper.ts

```diff
@@ -0,0 +1,341 @@
+              "Ensure your selection reflects the real time investment needed for the task.",
+            ].join("\n");
+
+            const content = formatContent(issue, metadata);
+            const timeLabel = timeLabels[0];
+            sourceStats.set(repoPath, (sourceStats.get(repoPath) || 0) + 1);
+
+            const messages = [
+              { role: "system", content: systemPrompt },
+              { role: "user", content },
+              { role: "assistant", content: timeLabel },
+            ];
+
+            allExamples.push({ messages, timeLabel, source: repoPath });
+            processedCount++;
+          }
+        }
+        console.log(`Processed ${processedCount} labeled issues from ${repoPath}`);
+      }
+    }
+
+    console.log("\nInitial Source Distribution:");
+    for (const [source, count] of sourceStats.entries()) {
+      console.log(`${source}: ${count} issues`);
+    }
+
+    if (allExamples.length === 0) {
+      throw new Error("No valid training data found!");
+    }
+
+    // Balance the dataset
+    console.log("\nBalancing dataset...");
+    const balancedExamples = balanceDataset(allExamples);
+    console.log(`\nFinal balanced dataset size: ${balancedExamples.length}`);
+
+    // Shuffle and split the balanced data
+    const trainSize = Math.floor(balancedExamples.length * TRAIN_SPLIT);
+    const trainData = balancedExamples.slice(0, trainSize);
+    const validationData = balancedExamples.slice(trainSize);
+
+    // Write training data
+    const trainJsonlData = trainData.map((data) => JSON.stringify({ messages: data.messages })).join("\n");
+    writeFileSync(trainOutputPath, trainJsonlData);
+    console.log(`\nTraining data (${trainData.length} examples) written to ${trainOutputPath}`);
+
+    // Write validation data
+    const validationJsonlData = validationData.map((data) => JSON.stringify({ messages: data.messages })).join("\n");
+    writeFileSync(validationOutputPath, validationJsonlData);
+    console.log(`\nValidation data (${validationData.length} examples) written to ${validationOutputPath}`);
+
+    // Log final counts
+    const labelCounts = new Map<string, number>();
+    for (const example of balancedExamples) {
+      labelCounts.set(example.timeLabel, (labelCounts.get(example.timeLabel) || 0) + 1);
+    }
+
+    console.log("\nFinal Label Distribution:");
+    for (const [label, count] of labelCounts.entries()) {
+      console.log(`${label}: ${count} examples`);
+    }
+
+    console.log(`\nTotal balanced examples: ${balancedExamples.length}`);
+    console.log(`Training examples: ${trainData.length}`);
+    console.log(`Validation examples: ${validationData.length}`);
+  } catch (error) {
+    console.error("Error processing issues:", error);
+    process.exit(1);
+  }
+}
+
+// Check if gh cli is installed and authenticated
+try {
+  // Use full path to gh binary for security
+  execSync("/usr/bin/gh auth status", { stdio: "ignore" });
```

Comment:
I wonder if there is a better direct api to access the `gh` program instead of going to the shell for execution.

Reply:
I'm not sure that is true as the cli uses the rest and graphql apis, that would have been cool tho. https://stackoverflow.com/questions/71443093/do-github-cli-commnds-use-rest-apis-of-github-is-there-any-rate-limit-for-using


Also you can use the `--paginate` flag with the cli and it should paginate for you

https://cli.github.com/manual/gh